File: LeetCodePractice/src/bst/PathSumII.java
package bst;

import java.util.*;

/*
Given the root of a binary tree and an integer targetSum, return all root-to-leaf paths where each path's sum equals targetSum.

A leaf is a node with no children.



Example 1:


Input: root = [5,4,8,11,null,13,4,7,2,null,null,5,1], targetSum = 22
Output: [[5,4,11,2],[5,8,4,5]]
Example 2:


Input: root = [1,2,3], targetSum = 5
Output: []
Example 3:

Input: root = [1,2], targetSum = 0
Output: []


Constraints:

The number of nodes in the tree is in the range [0, 5000].
-1000 <= Node.val <= 1000
-1000 <= targetSum <= 1000
 */
public class PathSumII {

	public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
        List<List<Integer>> res = new ArrayList<>();
        if(root == null){
            return res;
        }
        List<Integer> ls = new ArrayList<>();
        dfs(res, ls, root, targetSum);
        return res;
   }
   
   private void dfs(List<List<Integer>> res, List<Integer> ls, TreeNode root, int sum){
       if(root == null){
           return;
       }
       ls.add(root.val);
       if(root.val == sum && root.left == null && root.right == null){
           res.add(new ArrayList<>(ls));
       } else {
           dfs(res, ls, root.left, sum - root.val);
           dfs(res, ls, root.right, sum - root.val);  
       }
       ls.remove(ls.size() - 1); 
   }
}

File: LeetCodePython/PathSumII.py
"""
Given the root of a binary tree and an integer targetSum, return all root-to-leaf paths where the sum of the node values in the path equals targetSum. Each path should be returned as a list of the node values, not node references.

A root-to-leaf path is a path starting from the root and ending at any leaf node. A leaf is a node with no children.



Example 1:


Input: root = [5,4,8,11,null,13,4,7,2,null,null,5,1], targetSum = 22
Output: [[5,4,11,2],[5,8,4,5]]
Explanation: There are two paths whose sum equals targetSum:
5 + 4 + 11 + 2 = 22
5 + 8 + 4 + 5 = 22
Example 2:


Input: root = [1,2,3], targetSum = 5
Output: []
Example 3:

Input: root = [1,2], targetSum = 0
Output: []


Constraints:

The number of nodes in the tree is in the range [0, 5000].
-1000 <= Node.val <= 1000
-1000 <= targetSum <= 1000
"""
from typing import Optional, List


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class PathSumII:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> List[List[int]]:
        res = []
        if not root:
            return res

        def dfs(ls, cur, target):
            if not cur:
                return
            ls.append(cur.val)
            target -= cur.val
            if target == 0 and not cur.left and not cur.right:
                res.append(ls.copy())
            dfs(ls, cur.left, target)
            dfs(ls, cur.right, target)
            ls.pop()

        dfs([], root, targetSum)
        return res
